e3nn_jax/_src/s2grid.py

```diff
@@ -0,0 +1,354 @@
+    shb = jnp.einsum("lmj,bj,lmi,l->mbi", m, shb, m, n)  # [m, b, i]
+
+    size = coeffs.shape[:-1]
+    coeffs = coeffs.reshape(-1, coeffs.shape[-1])
+
+    # multiply spherical harmonics by their coefficients
+    signal_b = jnp.einsum("mbi,zi->zbm", shb, coeffs)  # [batch, beta, m]
+
+    signal = irfft(signal_b, res_alpha) * res_alpha
+    return signal.reshape(*size, *signal.shape[1:])
+
+
+def rfft(x: jnp.ndarray, l: int):
+    r"""Real fourier transform
+    Args:
+        x: `jnp.ndarray`
+            input array of shape ``(..., res_beta, res_alpha)``
+        l: int
+            value of `l` for which the transform is being run
+    Returns:
+        `jnp.ndarray`
+            transformed values - array of shape ``(..., res_beta, 2*l+1)``
+    """
+    x_reshaped = x.reshape((-1, x.shape[-1]))
+    x_transformed_c = jnp.fft.rfft(x_reshaped) # (..., 2*l+1)
+    x_transformed = jnp.concatenate([
+        jnp.flip(jnp.imag(x_transformed_c[..., 1:l+1]), -1) * -np.sqrt(2),
+        jnp.real(x_transformed_c[..., :1]),
+        jnp.real(x_transformed_c[..., 1:l+1]) * np.sqrt(2)
+    ], axis=-1)
+    return x_transformed.reshape((*x.shape[:-1], 2*l+1))
+
+
+def irfft(x: jnp.ndarray, res: int):
+    r"""Inverse of the real fourier transform
+    Args:
+        x: `jnp.ndarray`
+            array of shape ``(..., 2*l + 1)``
+        res: int
+            output resolution, has to be an odd number
+    Returns:
+        `jnp.ndarray`
+            positions on the sphere, array of shape ``(..., res, 3)``
+    """
+    assert res % 2 == 1
+
+    l = (x.shape[-1] - 1) // 2
+    x_reshaped = jnp.concatenate([
+        x[..., l:l+1],
+        (x[..., l+1:] + jnp.flip(x[..., :l], -1) * -1j) / np.sqrt(2),
+        jnp.zeros((*x.shape[:-1], l))
+    ], axis=-1).reshape((-1, x.shape[-1]))
+    x_transformed = jnp.fft.irfft(x_reshaped, res)
+    return x_transformed.reshape((*x.shape[:-1], x_transformed.shape[-1]))
+
+
+def _expand_matrix(ls):
+    """
+    conversion matrix between a flatten vector (L, m) like that
+    (0, 0) (1, -1) (1, 0) (1, 1) (2, -2) (2, -1) (2, 0) (2, 1) (2, 2)
+    and a bidimensional matrix representation like that
+                    (0, 0)
+            (1, -1) (1, 0) (1, 1)
+    (2, -2) (2, -1) (2, 0) (2, 1) (2, 2)
+    Args:
+        ls: list of l values
+    Returns:
+        tensor [l, m, l * m]
+    """
+    lmax = max(ls)
+    m = np.zeros((len(ls), 2 * lmax + 1, sum(2 * l + 1 for l in ls)))
+    i = 0
+    for j, l in enumerate(ls):
+        m[j, lmax - l : lmax + l + 1, i : i + 2 * l + 1] = np.eye(2 * l + 1)
+        i += 2 * l + 1
+    return jnp.asarray(m)
+
+
+def _rollout_sh(m, lmax):
+    """
+    Expand spherical harmonic representation.
+    Args:
+        m: jnp.ndarray of shape (..., (lmax+1)*(lmax+2)/2)
+    Returns:
+        jnp.ndarray of shape (..., (lmax+1)**2)
+    """
+    assert m.shape[-1] == (lmax+1) * (lmax+2) // 2
+    m_full = np.zeros((*m.shape[:-1], (lmax+1)**2))
```

Comment:
You should not convert the input that is a `jax` array into a `numpy` array.
You should use `jax.numpy` here and set the indices with the `x = x.at[i].set(y)` syntax.